src/Runner.php

```diff
@@ -11,15 +11,24 @@
 class Runner
 {
     protected $index;
+
+    /**
+     * @var MiddlewareStack
+     */
     protected $middleware;
 
-    public function run($middleware, ServerRequestInterface $request, ResponseInterface $response)
+    public function run(MiddlewareStack $middleware, ServerRequestInterface $request, ResponseInterface $response)
```

Comment:
Yes there is a required set of methods. However, typehinting against a concrete implementation requires that any changes be done as subclasses. If instead the typehint was against an interface someone could extend the library without inheritance.